system/lib/wasmfs/file_table.h

```diff
@@ -93,36 +93,44 @@ class FileTable {
     // return a reference, and can be used to check for the lack of an item
     // there without allocation (similar to how table[x] works on a JS object),
     // which keeps syntax concise.
-    struct Entry {
-      // Need to store a reference to the single global filetable, which is a
-      // local static variable.
-      Handle& fileTableHandle;
-      __wasi_fd_t fd;
+    class Entry;
 
-      operator std::shared_ptr<OpenFileState>() const;
+    Entry operator[](__wasi_fd_t fd);
 
-      Entry& operator=(std::shared_ptr<OpenFileState> ptr);
+    __wasi_fd_t add(std::shared_ptr<OpenFileState> openFileState);
+  };
+};
 
-      Entry& operator=(Entry& entry) {
-        return *this = std::shared_ptr<OpenFileState>(entry);
-      }
+class FileTable::Handle::Entry : public FileTable::Handle {
```

Comment:
Neat! I didn't know you could have out-of-line definitions of nested classes like this. WDYT about making `Entry` a sibling rather than child of `FileTable::Handle` so it would be `FileTable::Entry` rather than `FileTable::Handle::Entry`? Inheriting from `FileTable::Handle` still makes sense, but I think that would be just as clear and would make the full name shorter.

Reply:
I had to make it out of line, if we required Entry to be still nested in Handle. Perhaps making it a sibling makes more sense in this case.